Site Creation: error view

ScoutHarris opened the pull request:
Fixes #8555 

**To test:**

- Go through the site creation process.
- Trigger an error. 😄 
- @nate - let me know if want tips on triggering errors. There are two overall types: 
  - validation - created by a failure in `SiteCreationFields.validateFields`.
  - site create - created by a failure in any step in `SiteCreationService`.

- Verify the view looks like the 'implemented' image below.
- NOTE: The design has a 'Back' nav bar button. However, as discussed below, there is a 'Dismiss' button that exits the Site Creation process (i.e. dismisses the SC nav controller).
- The bold title will always be the same.
- The subtitle will change depending on the error triggered.

**Validation errors, messages and actions:**
NOTE: These are "sanity" validations. Since we do validation on each step prior to the site creation step, these will most likely never appear.
- site title: 
    - "The Site Title is missing."
    - 'Try again' goes to Site Details.
- domain: 
    - "The Site Domain is missing."
    - "The Site Domain contains wordpress.com."
    - 'Try again' goes to Domain Suggestion.
- theme: 
    - "The Site Theme is missing."
    - 'Try again' goes to Theme Selection.

**Site Creation errors, messages and actions:**
- validation: 
    - "The Site Domain is invalid."
    - 'Try again' goes to Domain Suggestion.
- getting default account: 
    - "We were unable to get your account information."
    - 'Try again' goes to Create Site, re-attempting site creation.
- creating site: 
    - "We were unable to create the site."
    - 'Try again' goes to Create Site, re-attempting site creation.
- setting tagline: 
    - "Your Site was created. Unfortunately, we were unable to set the Site Tagline. You can set the Tagline in the Site Settings, or try again now."
    - 'Try again' goes to Create Site, re-trying the process starting from setting the tagline.
- setting theme: 
    - "Your Site was created. Unfortunately, we were unable to set the Site Theme. You can set the Theme in the Site Settings, or try again now."
    - 'Try again' goes to Create Site, re-trying the process starting from setting the theme.
- syncing account: 
    - "Your Site was created. Unfortunately, we were unable to sync your account information. You can still access your site from My Sites. You can also retry syncing your account now."
    - 'Try again' goes to Create Site, re-trying the process starting from syncing the account.

**Implemented:**
![error_implemented](https://user-images.githubusercontent.com/1816888/35755870-3af3a172-0826-11e8-818f-b64ccfa584af.png)

**Designed:**
<img width="300" alt="error_designed" src="https://user-images.githubusercontent.com/1816888/35755873-4159355e-0826-11e8-8fb1-ac01b687c673.png">

ScoutHarris commented:
@nheagy , @folletto , @iamthomasbishop - I wasn't 100% sure about some of the actions and messages. So I'll take any and all feedback. Also, once we decide on those, I'll then add the 'Needs Copy Review' label for the final messages.

folletto commented:
> The button title and action will change depending on the error:
> Basically, if the failure is pre-site creation, the button text will be 'Try Again'.
> If the error is post-site creation, the button will be 'Dismiss'.

Is this referring to the various creation steps?

Ideally it should always be "Try Again" on the step it just failed, never restarting from the beginning.
See the Jan 30 discussion here: p77Llu-9vz-p2

...or maybe I'm just missing something here. :)

> "X is missing."

How is this possible? Is it just to be sure? In theory they shouldn't be able to go on if something is missing, right?

> Dismiss

The approach to use "Dismiss" instead of "Try Again" is interesting here, I wasn't considering it. It's a good idea (maybe to replicate on Android? @hypest can you review?).

The only advice here would be to however change the copy. Not "Dismiss" but "Skip and Proceed" or something analogous.

For everything else, I think we can proceed asking for a Copy Review when you're good.

hypest commented:
> > Dismiss
> 
> The approach to use "Dismiss" instead of "Try Again" is interesting here, I wasn't considering it. It's a good idea (maybe to replicate on Android? @hypest can you review?).

Thanks for the ping @folletto ! Not sure though which case exactly you'd like me to look at. Can you expand? Thanks!

folletto commented:
> Not sure though which case exactly you'd like me to look at. Can you expand? Thanks!

Can we do the same on Android? Can you cross-check we're doing the same thing on both platforms?
iOS and Android should have the same logic.

hypest commented:
Sure, I can try to mirror the design on android too after this PR gets finalized.

hypest commented:
> Can you cross-check we're doing the same thing on both platforms?
> iOS and Android should have the same logic.

I cross-checked with what we currently have implemented (and merged) for android @folletto . Here's the diff:

#### 1. Validation errors
We have no validation performed while on the "Creating/Error" screen. All validations have been performed in previous screens (e.g. can't even move past the Site Details screen if Site Title is missing) so, the "Retry" button doesn't apply here.
#### 2. Site Creation errors
* validation ("The Site Domain is invalid."). Not applicable since the domain is picked from the suggestions only, on a previous screen. We can add some validation in that previous screen but not sure what's valid and what's not. What's the check needed?
* getting default account. Not applicable for android. No extra network call is performed to get the currently logged in account info.
* creating site ("We were unable to create the site."). 
  * We're using the generic error messages as shown in #6929. Do you want us to specialize it?
  * "Try again" button already re-tries the site creation process so, no diff here.
* setting tagline ("Your Site was created. Unfortunately, we were unable to set the Site Tagline. You can set the Tagline in the Site Settings.").
  * We're using the generic error messages as shown in #6929. Do you want us to specialize it?
  * "Try again button" retries the process, starting from the set-tagline step. Should we revise to turn it into "Dismiss"?
* setting theme ("Your Site was created. Unfortunately, we were unable to set the Site Theme. You can set the Theme in the Site Settings.")
  * We're using the generic error messages as shown in #6929. Do you want us to specialize it?
  * "Try again button" retries the process, starting from the set-theme step. Should we revise to turn it into "Dismiss"?
* syncing account ("Your Site was created. Unfortunately, we were unable to sync your account information. You can still access your site from My Sites."
  * We're using the generic error messages as shown in #6929. Do you want us to specialize it?
  * Assuming we're talking about the second step in the process, right after the site is created remotely, the "Try again button" retries the process, starting from the fetch-site-information step. Should we revise to turn it into "Dismiss"?

Effectively, it seems that if we adjust this screen's android implementation to match the iOS one, the "Try again" button will only appear in one case, when the process failed in its start, and will be "Dismiss" in all other. We will need to undo the work done to implement the suggestions here: p77Llu-9vz-p2#comment-7920. So, to double check, should we go through with the undoing?

ScoutHarris commented:
@hypest , @folletto  - 

> All validations have been performed in previous screens (e.g. can't even move past the Site Details screen if Site Title is missing)

We do this as well. I should have been more clear about the validation errors here - they really should never appear as we do validations on each step. However, we do a final "sanity" validation check before creating the site, where the validation errors noted in this PR will be displayed. But again, the chance of that is minute.

ScoutHarris commented:
@hypest , @folletto -

There's a reason I started this PR with the comment 'I wasn't 100% sure about some of the actions'.  😄 

I believe I went wrong on the last three steps: setting tagline, setting theme, syncing account. Correct me if I misunderstood again, but the idea is if these occur we should retry the failed step on the newly created site (it seems so clear now...).  

So:
- Given that, there should never be a 'Dismiss' button - it should always be 'Try again'. (as I believe @hypest did initially)
- I'm unclear as to what the UI will be during these three 'Try again' states. Since we're not going back to the associated view (we're just trying again), do we re-display the Creating Site view?

nheagy commented on WordPress/Classes/ViewRelated/NUX/SiteCreationCreateSiteViewController.swift:
```diff
@@ -109,30 +151,119 @@ class SiteCreationCreateSiteViewController: NUXViewController {
+                setReturnViewController(for: .domainSuggestion)
+                return NSLocalizedString("The Site Domain is invalid.", comment: "Error shown during site creation process when the site domain validation fails.")
+            case .gettingDefaultAccount:
+                setReturnViewController(for: .createSite)
```
Heading back to the site creation VC isn't reattempting to create the site, but as we discussed it I think that's really easy to fix 😀

ScoutHarris commented on WordPress/Classes/Services/SiteCreationService.swift:
```diff
@@ -216,16 +221,15 @@ open class SiteCreationService: LocalCoreDataService {
-            let error = SiteCreationError.missingRESTAPI as NSError
+            let error = SiteCreationError.missingRESTAPI
             DDLogError("Error while creating site: Missing REST API.")
-            assertionFailure()
```
The assertions were causing the app to crash. 1) That seemed severe; 2) With the error view these issues are handled gracefully.

nheagy replied:
Yes, the app crashing is intentional, to surface these problems in development. More on this: these assertions only cause crashes in debug builds because the `NS_BLOCK_ASSERTIONS` macro is used for release builds.

We use these kinds of assertions throughout the app (as a project-wide search for `assertionfailure` will show). I think this is a bigger change than should be made as part of this PR; if you want to remove these I think it should be done as a separate PR, and that PR should cover more than the `SiteCreationService.swift`.

ScoutHarris replied:
Ahhhh! Thank you for the insight. It'll be like this never happened! 😉

ScoutHarris commented on WordPress/Classes/ViewRelated/Blog/NoResultsViewController.swift:
```diff
@@ -0,0 +1,103 @@
+
+    override var sourceTag: SupportSourceTag {
+        get {
+            if let parent = parent,
```
Ah, good point. Thank you sir.

ScoutHarris commented on WordPress/Classes/Services/SiteCreationService.swift:
```diff
@@ -147,10 +138,24 @@ open class SiteCreationService: LocalCoreDataService {
         }
 
         let validateBlogSuccessBlock = {
+
```
> Can you explain?

It made more sense logically to test for the default account just before it was used, instead of some seemingly random place.

> I'm a bit unsure if this would have implications for the old flow.

The `SignupService` has it's own `validateBlogSuccessBlock`, and `SignupService` doesn't call `createSite` directly (it calls the methods it needs directly) so there is no impact.

nheagy replied:
Got it—thanks!

PROCEED

folletto commented:
> So, to double check, should we go through with the undoing?

No, I just needed to be sure Android and iOS were aligned, and if not, that it was an explicit decision. Thanks for your review. :)

folletto commented:
> However, we do a final "sanity" validation check before creating the site, where the validation errors noted in this PR will be displayed. But again, the chance of that is minute.

I appreciate having double checks like this, even if they should never show up. :)

> the idea is if these occur we should retry the failed step on the newly created site (it seems so clear now...).

Correct! :)

> I'm unclear as to what the UI will be during these three 'Try again' states. Since we're not going back to the associated view (we're just trying again), do we re-display the Creating Site view?

Yes, ideally it should go back to the creating view, at the step actually being re-tried.

> It's conceivable that the user would be stuck and have no way out but to kill the app. 

Ideally... they should be able to "Back" out of it.
But as per the discussion above, we might want to be a bit clever to find a "Back" solution that is simple to implement. I didn't prescribe it in detail as I wanted to give you that flexibility not knowing what's simpler to build.

Among the options:

1. back always goes to the previous step, until the user is out
2. when back and retrying, should retry only what's left, not create a new site (this might be very complicated in terms of logic when going back, i.e. freezing already submitted fields)
3. Back becomes "Dismiss", and cancels everything.
4. Back becomes "Dismiss" and only shows up in case of error.

I think all of these have pros/cons, and I'm more keen in this first version to have a simple and maintainable solution that will allow iterating later, more than trying to be speculative for a perfect solution.

ScoutHarris commented:
@folletto , @nheagy  - regarding abandoning site creation on the error view, I vote for:

3. Back becomes "Dismiss", and cancels everything.

An error has already occurred, that's why they're seeing this page. 😄  But we can simply add a 'Dismiss' button so they can bail at will.

folletto commented:
Ok for me. :)

ScoutHarris commented:
@nheagy , @folletto  - I've updated the following per discussions. This has been noted in the PR description, but for reference:

Added a navigation bar 'Dismiss' button that exits the Site Creation process (i.e. dismisses the Site Creation navigation controller).

Corrected these actions, with updated messaging:
- setting tagline: 
    - "Your Site was created. Unfortunately, we were unable to set the Site Tagline. You can set the Tagline in the Site Settings, or try again now."
    - 'Try again' goes to Create Site, re-trying the process starting from setting the tagline.
- setting theme: 
    - "Your Site was created. Unfortunately, we were unable to set the Site Theme. You can set the Theme in the Site Settings, or try again now."
    - 'Try again' goes to Create Site, re-trying the process starting from setting the theme.
- syncing account: 
    - "Your Site was created. Unfortunately, we were unable to sync your account information. You can still access your site from My Sites. You can also retry syncing your account now."
    - 'Try again' goes to Create Site, re-trying the process starting from syncing the account.

And finally - @nheagy - ready for another review please!

benhuberman commented:
@ScoutHarris on the copy level, a couple of quick notes:
- "Your site was created" -- could we switch to the active voice here? It seems like a celebratory moment that's undercut somewhat by the lack of agency. I'd consider either "We've successfully created your site" or "Your new site is now live" or "Congratulations on creating a new site!" (or something along those lines).

> we were unable to set the Site Tagline
> we were unable to set the Site Theme
Here, "site tagline" and "site theme" are used generically (I think -- correct me if I'm wrong!) rather than as UI headings, so I'd vote to keep them in sentence case, e.g, "We were unable to set your site's tagline" "We were unable to set our site's theme," etc.

> set the Tagline in the Site Settings
> set the Theme in the Site Settings
Again, I don't think there's a need to capitalize either "Tagline" or "Theme," and it might also read smoother if you tweak "in the Site Settings" to something like "in the Site Settings screen." Here, keeping it title case makes sense as long as you make it clear you're referring to a UI heading.

ScoutHarris commented:
@benhuberman - thank you for the review!

> "Your site was created" -- could we switch to the active voice here?

I went with "We've successfully created your site." Since this in an error situation, I didn't want to be _too_ excited here. We have a much more exciting page when site creation is 100% successful. 😁 

> "site tagline" and "site theme" are used generically

You are correct. They are now lower case.

benhuberman commented:
> Since this in an error situation, I didn't want to be too excited here. 
Good point!

nheagy commented:
I think this looks good 😀 

:shipit: 

Nice work @ScoutHarris
